storage/engine/mvcc.go

```diff
@@ -653,54 +653,47 @@ func mvccPutInternal(engine Engine, ms *proto.MVCCStats, key proto.Key, timestam
-		if meta.Txn != nil && (txn == nil || !bytes.Equal(meta.Txn.ID, txn.ID)) {
-			return &proto.WriteIntentError{Intents: []proto.WriteIntentError_Intent{{Key: key, Txn: *meta.Txn}}}
-		}
+		// There is existing metadata for this key; ensure our write is permitted.
 
-		// We can update the current metadata only if both the timestamp
-		// and epoch of the new intent are greater than or equal to
-		// existing. If either of these conditions doesn't hold, it's
-		// likely the case that an older RPC is arriving out of order.
-		//
-		// Note that if meta.Txn!=nil and txn==nil, a WriteIntentError was
-		// returned above.
-		if !timestamp.Less(meta.Timestamp) &&
-			(meta.Txn == nil || txn.Epoch >= meta.Txn.Epoch) {
-			// If this is an intent and timestamps have changed,
-			// need to remove old version.
-			if meta.Txn != nil && !timestamp.Equal(meta.Timestamp) {
+		if meta.Txn != nil {
+			// There is an uncommitted write intent.
+
+			if txn == nil || !bytes.Equal(meta.Txn.ID, txn.ID) {
+				// The current Put operation does not come from the same
+				// transaction.  This should not happen since range should check
+				// the existing write intent before executing any Put action at
+				// MVCC level.
+				return &proto.WriteIntentError{Intents: []proto.WriteIntentError_Intent{{Key: key, Txn: *meta.Txn}}}
+			} else if txn.Epoch < meta.Txn.Epoch || timestamp.Less(meta.Timestamp) {
+				// Ignore old writes from the current transaction; it's just an
+				// RPC arriving out of order.
+				return nil
+			} else if !timestamp.Equal(meta.Timestamp) {
+				// We are replacing our own write intent.
 				versionKey := mvccEncodeTimestamp(metaKey, meta.Timestamp)
 				if err := engine.Clear(versionKey); err != nil {
 					return err
 				}
 			}
-			newMeta = &buf.newMeta
-			buf.newMeta = proto.MVCCMetadata{Txn: txn, Timestamp: timestamp}
-		} else if timestamp.Less(meta.Timestamp) && meta.Txn == nil {
-			// If we receive a Put request to write before an already-
-			// committed version, send write tool old error.
-			return &proto.WriteTooOldError{Timestamp: timestamp, ExistingTimestamp: meta.Timestamp}
 		} else {
-			// Otherwise, it's an old write to the current transaction. Just ignore.
-			return nil
+			// No pending transaction, so we can write as long as it is not
+			// in the past.
+			if !meta.Timestamp.Less(timestamp) {
+				//if timestamp.Less(meta.Timestamp) {
+				return &proto.WriteTooOldError{Timestamp: timestamp, ExistingTimestamp: meta.Timestamp}
+			}
 		}
-	} else { // In case the key metadata does not exist yet.
-		// If this is a delete, do nothing!
+	} else {
+		// No existing metadata record. If this is a delete, do nothing;
+		// otherwise we can perform the write.
 		if value.Deleted {
 			return nil
 		}
-		// Create key metadata.
 		meta = nil
-		newMeta = &buf.newMeta
-		buf.newMeta = proto.MVCCMetadata{Txn: txn, Timestamp: timestamp}
 	}
+	newMeta := &buf.newMeta
```

Comment:
`meta` stays nil if it was set on line 696, but if we took the other half of the `if ok` it will be non-nil. This is weird; it should probably be the other way around (`meta` only becomes non-nil if `ok` is true).